flumine/markets/blotter.py

```diff
@@ -103,6 +104,25 @@ def process_cleared_orders(self, cleared_orders) -> list:
 
     """ position """
 
+    def market_exposure(self, strategy, num_runners, num_winners: int = 1) -> float:
```

Comment:
I have tried to keep all order / exposure calculations in the blotter object, if we were to move something to the Market then I think everything needs to move. Don't like the idea of these functions spread around the codebase.

I guess the only alternative is to pass the market to the blotter...

Reply:
Understood. Big fan of consistency.

Considered an alternative: namely receiving still market as a parameter on the market_exposure method, but not casting it to the Market class, since that's what causes the circular reference. But it felt a bit icky, and became so when I realises that I really ought to handle the case where a user passes in a different market to the one for which the blotter instance is an attribute.

So ensuring that we're dealing with the same market by passing an instance of it into the blotter constructor looks like the neatest solution, but it's a bit of a sledgehammer with tweaking needed to a lot of tests to pass a mock market object in place of a market_id str.

I'm happy to have a run at that, but would you like some more time to ponder the implications first?

Reply:
Let's have a think, I feel like we are missing a simple refactor for this to work nicely...

Reply:
Not really a refactor, but one possibility would be to accept the num_runners and num_winners as optional parameters on the blotter constructor, and then pass them in when the blotter is instantiated in the market class. That would be minimally invasive, i.e. would allow them to be removed from the market_exposure method call and would avoid the re-writing of existing tests (other than those that are part of this PR).

Reply:
num of runners can change though

Reply:
I am thinking we just get this in as is, see how people use it, can always update later

Reply:
Removing the parameters does seem to be more trouble than it's worth, and best to move on from at the moment.

Per your earlier question, I'm not currently seeing any reason why it shouldn't be possible to calculate market exposure across all strategies, i.e. handling "None" as the strategy parameter. I was planning to add some tests to validate this. Should I move forward on this (as this is built on the selection_exposure method, it would likely mean to that method and its tests too)? Or would you rather go with it in its current form.

Reply:
Lets leave it as is, will get this released this morning 😄

Reply:
> We could abstract it by:
> 
> ```python
> def market_exposure(self, strategy, market: Market) -> float:
>     num_runners = market.market_book.number_of_runners
>     num_winners = market.market_book.number_of_winners
> ```
> 
> Then becomes much easier to use as a user, as you just chuck it the market object and let the function do the horrible stuff.

I know Peter said that passing the market is a circular reference, but what about this?
```python
def market_exposure(self, strategy, market_book) -> float:
     num_runners = market_book.number_of_runners
     num_winners = market_book.number_of_winners
 ```

Reply:
That's a good idea that neatly side-steps the circular reference, but it would be a breaking change.

I have the revised and tested code that would implement this, so will do a pull request, but it will be for Liam to decide whether or not he wants to merge it.

Reply:
https://github.com/liampauling/flumine/pull/544

Reply:
Change works for me 👍